File: lib/api/invitations.ts
// Production Phase 4B organization team invitations client.
//
// An org admin can invite, list, and revoke invitations; a signed-in user can
// accept one with a token. Invitation tokens are delivered by email in
// production; outside production the backend also returns a dev token so the
// flow can be completed locally. The token is never logged.

import { API_BASE_URL, authHeaders, safeFetch } from "./client";

export type Invitation = {
  invitationId: string;
  organizationId: string;
  email: string;
  role: string;
  status: string;
  invitedByUserId: string | null;
  acceptedByUserId: string | null;
  expiresAt: string | null;
  acceptedAt: string | null;
  revokedAt: string | null;
  createdAt: string | null;
};

export type InvitationLookup = {
  organizationId: string;
  organizationName: string | null;
  email: string;
  role: string;
  status: string;
  acceptable: boolean;
  expiresAt: string | null;
};

export type MutationResult<T> = {
  ok: boolean;
  backendReachable: boolean;
  data?: T;
  error?: string;
};

function mapInvitation(raw: Record<string, unknown>): Invitation {
  return {
    invitationId: raw.invitation_id as string,
    organizationId: raw.organization_id as string,
    email: raw.email as string,
    role: raw.role as string,
    status: raw.status as string,
    invitedByUserId: (raw.invited_by_user_id as string) ?? null,
    acceptedByUserId: (raw.accepted_by_user_id as string) ?? null,
    expiresAt: (raw.expires_at as string) ?? null,
    acceptedAt: (raw.accepted_at as string) ?? null,
    revokedAt: (raw.revoked_at as string) ?? null,
    createdAt: (raw.created_at as string) ?? null,
  };
}

async function mutate<T>(
  path: string,
  body: unknown,
  mapper: (raw: Record<string, unknown>) => T,
): Promise<MutationResult<T>> {
  try {
    const res = await fetch(`${API_BASE_URL}${path}`, {
      method: "POST",
      headers: authHeaders({ "Content-Type": "application/json" }),
      body: body === undefined ? undefined : JSON.stringify(body),
      cache: "no-store",
    });
    if (!res.ok) {
      let detail = `Request failed (${res.status}).`;
      try {
        const parsed = (await res.json()) as { detail?: string };
        if (parsed.detail) detail = parsed.detail;
      } catch {
        // Keep the generic message.
      }
      return { ok: false, backendReachable: true, error: detail };
    }
    return {
      ok: true,
      backendReachable: true,
      data: mapper((await res.json()) as Record<string, unknown>),
    };
  } catch {
    return {
      ok: false,
      backendReachable: false,
      error: "Backend unavailable. Start the API and try again.",
    };
  }
}

export async function listInvitations(
  organizationId: string,
): Promise<Invitation[] | null> {
  const data = await safeFetch<Record<string, unknown>[]>(
    `/api/v1/organizations/${organizationId}/invitations`,
  );
  return data ? data.map(mapInvitation) : null;
}

export type CreateInvitationResult = {
  invitation: Invitation;
  devInviteToken: string | null;
  emailSent: boolean;
};

export async function createInvitation(
  organizationId: string,
  payload: { email: string; role: string },
): Promise<MutationResult<CreateInvitationResult>> {
  return mutate<CreateInvitationResult>(
    `/api/v1/organizations/${organizationId}/invitations`,
    { email: payload.email, role: payload.role },
    (raw) => ({
      invitation: mapInvitation(raw.invitation as Record<string, unknown>),
      devInviteToken: (raw.dev_invite_token as string) ?? null,
      emailSent: (raw.email_sent as boolean) ?? false,
    }),
  );
}

export async function revokeInvitation(
  organizationId: string,
  invitationId: string,
): Promise<MutationResult<Invitation>> {
  return mutate<Invitation>(
    `/api/v1/organizations/${organizationId}/invitations/${invitationId}/revoke`,
    undefined,
    mapInvitation,
  );
}

export async function lookupInvitation(
  token: string,
): Promise<InvitationLookup | null> {
  const data = await safeFetch<Record<string, unknown>>(
    `/api/v1/invitations/lookup?token=${encodeURIComponent(token)}`,
  );
  if (!data) return null;
  return {
    organizationId: data.organization_id as string,
    organizationName: (data.organization_name as string) ?? null,
    email: data.email as string,
    role: data.role as string,
    status: data.status as string,
    acceptable: (data.acceptable as boolean) ?? false,
    expiresAt: (data.expires_at as string) ?? null,
  };
}

export async function acceptInvitation(
  token: string,
): Promise<MutationResult<{ organizationId: string; role: string; detail: string }>> {
  return mutate(
    "/api/v1/invitations/accept",
    { token },
    (raw) => ({
      organizationId: raw.organization_id as string,
      role: raw.role as string,
      detail: raw.detail as string,
    }),
  );
}

File: backend/app/api/v1/invitations.py
"""Organization team invitation API routes (Production Phase 4B).

An org_admin can invite a teammate by email with a role, list invitations, and
revoke a pending invitation. A signed-in user can accept an invitation with the
token, which creates an active membership. A public lookup returns a safe preview
for the accept page. Invitation tokens are never echoed except as a dev token
outside production.

These routes control team membership and audit attribution only. They never grant
engineering authority and never imply approval, certification, or compliance.
"""

from __future__ import annotations

from fastapi import APIRouter, Depends, Query
from sqlalchemy.orm import Session

from app.core.config import get_settings
from app.db import models
from app.db.database import get_db
from app.schemas.invitation import (
    InvitationAcceptRequest,
    InvitationAcceptResponse,
    InvitationCreateRequest,
    InvitationCreateResponse,
    InvitationLookupResponse,
    InvitationResponse,
)
from app.services import access_control_service, invitation_service, mailer
from app.services.access_control_service import (
    get_current_user,
    get_optional_user,
    require_org_admin,
)

router = APIRouter(tags=["invitations"])


@router.post(
    "/organizations/{organization_id}/invitations",
    response_model=InvitationCreateResponse,
    status_code=201,
)
def create_invitation(
    organization_id: str,
    body: InvitationCreateRequest,
    user: models.UserAccount | None = Depends(get_optional_user),
    db: Session = Depends(get_db),
) -> InvitationCreateResponse:
    """Invite a teammate to the organization. Requires org_admin."""

    admin = require_org_admin(db, organization_id, user)
    record, token = invitation_service.create_invitation(
        db,
        organization_id=organization_id,
        email=body.email,
        role=body.role,
        invited_by_user_id=admin.user_id,
    )
    db.commit()
    delivery = mailer.send_email(
        to=record.email,
        category="team_invitation",
        subject="You're invited to a Civil Engineer AI workspace",
        body=(
            "You have been invited to join a workspace. Use this token to "
            f"accept the invitation. Token: {token}"
        ),
    )
    settings = get_settings()
    return InvitationCreateResponse(
        invitation=InvitationResponse(
            **invitation_service.invitation_public_dict(record)
        ),
        dev_invite_token=token if settings.expose_dev_tokens else None,
        email_sent=bool(delivery.get("sent")),
    )


@router.get(
    "/organizations/{organization_id}/invitations",
    response_model=list[InvitationResponse],
)
def list_invitations(
    organization_id: str,
    user: models.UserAccount | None = Depends(get_optional_user),
    db: Session = Depends(get_db),
) -> list[InvitationResponse]:
    """List the organization's invitations. Requires org_admin."""

    require_org_admin(db, organization_id, user)
    return [
        InvitationResponse(**invitation_service.invitation_public_dict(inv))
        for inv in invitation_service.list_invitations(db, organization_id)
    ]


@router.post(
    "/organizations/{organization_id}/invitations/{invitation_id}/revoke",
    response_model=InvitationResponse,
)
def revoke_invitation(
    organization_id: str,
    invitation_id: str,
    user: models.UserAccount | None = Depends(get_optional_user),
    db: Session = Depends(get_db),
) -> InvitationResponse:
    """Revoke a pending invitation. Requires org_admin."""

    require_org_admin(db, organization_id, user)
    record = invitation_service.revoke_invitation(
        db, organization_id=organization_id, invitation_id=invitation_id
    )
    db.commit()
    return InvitationResponse(
        **invitation_service.invitation_public_dict(record)
    )


@router.get("/invitations/lookup", response_model=InvitationLookupResponse)
def lookup_invitation(
    token: str = Query(...), db: Session = Depends(get_db)
) -> InvitationLookupResponse:
    """Return a safe preview of an invitation for the accept page (public)."""

    return InvitationLookupResponse(
        **invitation_service.lookup_invitation(db, token)
    )


@router.post("/invitations/accept", response_model=InvitationAcceptResponse)
def accept_invitation(
    body: InvitationAcceptRequest,
    user: models.UserAccount = Depends(get_current_user),
    db: Session = Depends(get_db),
) -> InvitationAcceptResponse:
    """Accept an invitation for the signed-in user, creating a membership."""

    membership = invitation_service.accept_invitation(
        db, token=body.token, user=user
    )
    db.commit()
    org = access_control_service.get_organization(
        db, membership.organization_id
    )
    return InvitationAcceptResponse(
        organization_id=membership.organization_id,
        role=membership.role,
        membership_id=membership.membership_id,
        detail=f"You have joined {org.organization_name}.",
    )
